### ISSUE TITLE
Rho problems: Type checking for constraints parameter in higher rank type not working properly

### ISSUE THREAD

**Issue by yuxuanchiadm** [Frege/frege#336]
Test code that should not pass type checking attached:
```
foo :: (forall a. Something a => a -> a) -> (forall b. b -> b)
foo = id
-- Code below type checked properly and refused by compiler
-- foo f b = f b

class Something a

instance Something String

main = do
  println $ foo id $ "bar"
```
This will generate code like this:
```
public interface CSomething<𝓐>  {
  
}
final public static class ISomething_String implements CSomething<String/*<Character>*/> {
  public ISomething_String() {}
  final public static ISomething_String it = new ISomething_String();
}

final public static <𝓑> 𝓑 foo(final Func.U<CSomething<Object>, Func.U<Object, Object>> arg$1, final Lazy<𝓑> arg$2) {
  return Func.<Func.U<𝓑, 𝓑>, Func.U<𝓑, 𝓑>>coerceU(Category.ICategory_$minus$gt.it.<Func.U<𝓑, 𝓑>>ƒid())
      .apply(RunTM.<Func.U<𝓑, 𝓑>>cast(arg$1)).call().apply(arg$2).call();
}
```

Maybe related to #277.

**Comment by yuxuanchiadm** [comment]
This will generate code like this:
```
final public class Issue227  {
  

public interface CSomething<𝓐>  {
  
}
final public static class ISomething_String implements CSomething<String/*<Character>*/> {
  public ISomething_String() {}
  final public static ISomething_String it = new ISomething_String();
}

final public static <𝓐> 𝓐 foo(final Func.U<CSomething<Object>, Func.U<Object, Object>> arg$1, final Lazy<𝓐> arg$2) {
  return Func.<Func.U<𝓐, 𝓐>, Func.U<𝓐, 𝓐>>coerceU(Category.ICategory_$minus$gt.it.<Func.U<𝓐, 𝓐>>ƒid())
      .apply(RunTM.<Func.U<𝓐, 𝓐>>cast(arg$1)).call().apply(arg$2).call();
```

**Comment by yuxuanchiadm** [comment]
So after some code digging. I just found we are trying to apply type ```(𝓑 -> 𝓑)``` to type variable of function ```id```. That's fine because we will eventually need ```id :: (𝓑 -> 𝓑) -> (𝓑 -> 𝓑)```. But we just forget about we need to apply constraint singleton object to ```arg$1``` first.
TL;DR: We just try coerce ```forall a. Something a => a -> a``` to ```b -> b``` but forget about to deal with constraint thing.

**Comment by yuxuanchiadm** [comment]
Looks like something is wrong inside checkConstraints in Typecheck.fr. We got:
```
id :: a -> a
foo :: (forall a. Something a => a -> a) -> (forall b. b -> b)

sigma           forall b. (forall a. Something a => a -> a) -> b -> b
  rho           (forall a. Something a => a -> a) -> b -> b
    sigma       forall a. Something a => a -> a
      rho       Something a => a -> a
        sigma   a
          rho   a
            tau a
        rho     a
          tau   a
    rho         b -> b
      sigma     b
        rho     b
          tau   b
      rho       b
        tau     b
```
That should failed to check which isn't

**Comment by Ingo60** [comment]
Strange, indeed.
I'm not yet sure whether the error is in the code generation or in typechecking, though.

**Comment by yuxuanchiadm** [comment]
Should be in type checking imo. I was thinking this is code generation problem before too. But that should fail first in type checking so compiler never goes to code generation.

**Comment by yuxuanchiadm** [comment]
So after some nightmare debugging and code tracking:
```
-- ety
rho          (forall a. a -> a) -> b -> b
  sigma      forall a. a -> a
    rho      a -> a
      sigma  a
        rho  a
         tau a
      rho    a
        tau  a
  rho        b -> b
    sigma    b
      rho    b
        tau  b
    rho      b
      tau    b

-- x.typ
sigma   a -> a
  rho   a -> a
    tau a -> a
```
We got this when we finally do type checking in checkConstraints. The type we got in ```ety``` is not quite right. Leak some constraints things

**Comment by yuxuanchiadm** [comment]
Actually the type checking part of ```checkConstraints``` looks will not work for higher rank type too. It seems does nothing to higher rank type.

**Comment by Ingo60** [comment]
Here is another example:

    foo :: (forall a. Something a => a -> a) -> [b -> b]
    foo = (:[])
    -- Code below type checked properly and refused by compiler
    -- foo f b = f b

    -- for easier tracing, we dont use standard id which has a more complex type
    identity x = x

    idx = identity

    class Something a

    instance Something String

    main = do
       println  ( head (foo identity)  "bar" )

Turns out type checking works ok as long as the higher kinded argument is named somewhere in the right hand side. OTOH, the definition gets eta-expanded after type checking and before code generation.

So an easy fix would be to make sure that higher kinded arguments get named and silently eta expand the definition when they are not. 

The price would be maybe a silly error message. For example, if someone wrote 

    foo = true

The message would normally be:

    type error in expression true 
    type is apparently Bool 
    does not match function type t1→t2

So, if we would silently change this to 

    foo f = true f

we would get some more confusing messages. But we could prevent this by just type checking as now, and then again  if eta expansion on higher kinded arguments ist possible **unless** there had been any errors already. Then in the case 

     foo = pure

we would just get 

    inferred type is more constrained than expected type 
    inferred: Something b ⇒ (∀ a.Something a ⇒ a  → a) → [b→b] 
    expected: (∀ a.a → a) → [b→b]

(I'm not quite sure why it tells a wrong type that is expected, but ok)

**Comment by Ingo60** [comment]
Another issue is, even if one corrects the type signature, it still generates invalid code that gets a class cast exception on runtime. 
But if you eta-expand the definition, it all works.

So I would think eta expanding such functions during type checking is the right thing to do.

**Comment by yuxuanchiadm** [comment]
I think the reason of we got runtime exception is not code generation problem. It simply means type checker missed some invalid code that should refused by type checking if type checker works as intended.
```
foo :: (forall a. Something a => a -> a) -> [b -> b]
foo = (:[])
```
The intended compiler behavior of code above is type checker refuse it. And therefore it should never goes to java code generation phase.

**Comment by Ingo60** [comment]
Here's the reason why without eta expansion the type check must fail:

What happens during type checking is essentially that the type checker reproduces the RHS giving each term a type. (This information is used later in code generation.)

The type checking of type class  constraints is done by collecting the open constraints in the typed RHS. And here's the point: a term that doesn't appear cannot be assigned a type at all, so there is nothing to collect. But it is specifically the higher kinded function passed to foo that needs to get instantiated (i.e. applied a context-object).

For example, with

     foo :: Some b => (forall a. Some a. a -> a) -> b -> b    -- I prefer to write it thus
     foo x = id x
 
we get ($1 denote type variables):

      id x :: (b -> b)
      id :: $1 -> $1                (instantiated forall a. a->a)
      $1 = (b -> b)
      x :: Some $2 => $2 -> $2      (instantiated)
      x :: $1
      $1 = ($2 -> $2)
      (b->b) = ($2 -> $2)
      b = $2

So, in the end we have the RHS fully typed:

      (id :: (b -> b) -> (b -> b)) (x :: Some b => b -> b) :: b -> b

Here, the constraints collector finds `Some b` and complains that the type is more constrainted than the given type. But now, do the same game with the point free Version, and you'll see that this `Some b` appears nowhere. Of course not: There is just no term who gives rise to this. But this can only happen with constrained higher kinded function arguments, all other constraints come from mentioning a function that is constrained.

**Comment by yuxuanchiadm** [comment]
Looks like we also have some issues on Rank-N types. Compiled fine. Runtime exception. Test code attached:
```
class Something a

instance Something String

identity :: a -> a
identity a = a

foo :: (forall a. Something a => (forall b. Something b => (forall c. Something c => c -> c) -> b -> b) -> a -> a)
foo f a = f identity a

main = println (foo identity "ABC")
```
